airflow/contrib/hooks/spark_submit_hook.py

```diff
@@ -262,16 +317,82 @@ def _process_log(self, itr):
+                match_driver_id = re.search('(driver-[0-9\-]+)', line)
+                if match_driver_id:
+                    self._driver_id = match_driver_id.groups()[0]
+                    self.log.info("identified spark driver id: {}".format(self._driver_id))
+
+            self.log.debug("spark submit log: {}".format(line))
+
+    def _process_spark_status_log(self, itr):
+        """
+        parses the logs of the spark driver status query process
+
+        :param itr: An iterator which iterates over the input of the subprocess
+        """
+        # Consume the iterator
+        for line in itr:
+            line = line.strip()
+
+            # Check if the log line is about the driver status and extract the status.
+            if "driverState" in line:
+                self._driver_status = line.split(' : ')[1].replace(',', '').replace('\"', '').strip()
+
+            self.log.debug("spark driver status log: {}".format(line))
+
+    def _start_driver_status_tracking(self):
+        """
+        Polls the driver based on self._driver_id to get the status.
+        Finish successfully when the status is FINISHED.
+        Finish failed when the status is ERROR/UNKNOWN/KILLED/FAILED.
+
+        Possible status:
+            SUBMITTED: Submitted but not yet scheduled on a worker
+            RUNNING: Has been allocated to a worker to run
+            FINISHED: Previously ran and exited cleanly
+            RELAUNCHING: Exited non-zero or due to worker failure, but has not yet started running again
+            UNKNOWN: The status of the driver is temporarily not known due to master failure recovery
+            KILLED: A user manually killed this driver
+            FAILED: The driver exited non-zero and was not supervised
+            ERROR: Unable to run or restart due to an unrecoverable error (e.g. missing jar file)
+        """
+        # Keep polling as long as the driver is processing
+        while self._driver_status not in ["FINISHED", "UNKNOWN", "KILLED", "FAILED", "ERROR"]:
+
+            self.log.debug("polling status of spark driver with id {}".format(self._driver_id))
+
+            poll_drive_status_cmd = self._build_track_driver_status_command()
+            status_process = subprocess.Popen(poll_drive_status_cmd,
+                                              stdout=subprocess.PIPE,
+                                              stderr=subprocess.STDOUT,
+                                              bufsize=-1,
+                                              universal_newlines=True)
+
+            self._process_spark_status_log(iter(status_process.stdout.readline, ''))
+            returncode = status_process.wait()
+
+            if returncode:
+                raise AirflowException(
+                    "Failed to poll for the driver status: returncode = {}".format(returncode)
+                    )
+
+            # Sleep for 1 second as we do not want to spam the cluster
```

Comment:
If we move this sleep at the beginning of the `while` loop, we'll avoid waiting one last time after the batch is finished => we'll be slightly faster end-to-end :)